File: plugins/rust/cli/src/cli_replay.rs
use lotto::base::*;
use lotto::cli::flags::*;
use lotto::cli::*;
use lotto::engine::flags::*;
use lotto::*;
use rinflex::handlers::flags::*;

#[subcmd(
    name="replay-rs",
    args="",
    desc="replay in rust",
    engine_flags=true,
    cli_flags=[
        FLAG_INPUT.get(), FLAG_OUTPUT.get(), FLAG_VERBOSE.get(), FLAG_REPLAY_GOAL.get(),
        FLAG_TEMPORARY_DIRECTORY.get(), FLAG_NO_PRELOAD.get(), FLAG_LOGGER_BLOCK.get(), FLAG_BEFORE_RUN.get(), FLAG_AFTER_RUN.get()
    ],
    defaults=|flags: &mut Flags| {
        flags.set_default(&flag_stable_address_method(),
                          Value::U64(StableAddressMethod::STABLE_ADDRESS_METHOD_MAP as u64));
        flags.set_default(&FLAG_HANDLER_ADDRESS_ENABLED, Value::Bool(true));
        flags.set_default(&FLAG_HANDLER_CAS_ENABLED, Value::Bool(true));
        flags.set_default(&FLAG_HANDLER_EVENT_ENABLED, Value::Bool(true));
        flags.set_default(&FLAG_HANDLER_STACKTRACE_ENABLED, Value::Bool(true));
    },
    group=raw::subcmd_group::SUBCMD_GROUP_TRACE
)]
fn replay(_args: &mut Args, flags: &mut Flags) -> SubCmdResult {
    envvar_set! {
        "LOTTO_LOGGER_FILE" => flags.get_sval(&FLAG_LOGGER_FILE),
    };

    println!("trace file: {}", flags.get_sval(&FLAG_INPUT));

    let last_clk = cli_trace_last_clk(flags.get_sval(&FLAG_INPUT));
    let goal = flags.get_uval(&FLAG_REPLAY_GOAL);
    if goal > last_clk {
        flags.set_default(&FLAG_REPLAY_GOAL, Value::U64(last_clk));
    }

    let input_fn = flags.get_sval(&FLAG_INPUT);
    let mut rec = Trace::load_file(input_fn);
    let first: &Record = match rec.next(raw::record::RECORD_START) {
        None => {
            eprintln!("error: empty trace file '{}'", input_fn);
            return Err(1.into());
        }
        Some(first) => first,
    };

    let args = first.args().to_owned();
    rec.advance();
    rec.next(raw::record::RECORD_CONFIG).unwrap().unmarshal();
    drop(rec);

    envvar_set! {
        "LOTTO_REPLAY" => flags.get_sval(&FLAG_INPUT),
        "LOTTO_LOGGER_BLOCK" => flags.get_sval(&FLAG_LOGGER_BLOCK),
    };

    let output_fn = flags.get_sval(&FLAG_OUTPUT);
    envvar_set! {
        "LOTTO_RECORD" => output_fn,
    };

    let record_granularity: RecordGranularities = flags
        .get_uval(unsafe { &raw::flag_record_granularity() })
        .into();
    envvar_set! {
        "LOTTO_RECORD_GRANULARITY" => record_granularity.to_string()
    };

    println!("[lotto] replaying: {}", flags.get_sval(&FLAG_INPUT));
    println!("[lotto] recording: {}", flags.get_sval(&FLAG_OUTPUT));

    preload(
        flags.get_sval(&FLAG_TEMPORARY_DIRECTORY),
        flags.is_on(&FLAG_VERBOSE),
        !flags.is_on(&FLAG_NO_PRELOAD),
        flags.get_sval(&flag_memmgr_runtime()),
        flags.get_sval(&flag_memmgr_user()),
    );

    adjust(flags.get_sval(&FLAG_INPUT));
    let exitcode = execute(&args, flags, true);

    if cfg!(feature = "stable_address_map") {
        use rinflex::handlers::order_enforcer;
        order_enforcer::cli_reset();
        adjust(flags.get_sval(&FLAG_OUTPUT));
        if order_enforcer::should_discard() {
            println!(
                "number of constraints: {}",
                order_enforcer::HANDLER.fin.constraints.len()
            );
            println!(
                "should_discard = {}",
                order_enforcer::HANDLER.fin.should_discard
            );
            println!("** INVALID **");
        } else {
            println!("valid");
        }
    }
    SubCmdResult::from_error_code(exitcode)
}

File: src/driver/commands/cli_replay.c
/*******************************************************************************
 * replay
 ******************************************************************************/
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <lotto/base/envvar.h>
#include <lotto/base/record_granularity.h>
#include <lotto/driver/exec.h>
#include <lotto/driver/exec_info.h>
#include <lotto/driver/flagmgr.h>
#include <lotto/driver/flags/memmgr.h>
#include <lotto/driver/flags/sequencer.h>
#include <lotto/driver/preload.h>
#include <lotto/driver/record.h>
#include <lotto/driver/subcmd.h>
#include <lotto/driver/trace.h>
#include <lotto/driver/utils.h>
#include <lotto/engine/pubsub.h>
#include <lotto/engine/statemgr.h>
#include <lotto/sys/stdio.h>
#include <lotto/sys/stdlib.h>

int
replay(args_t *args, flags_t *flags)
{
    uint64_t verbose = flag_verbose_count(flags);

    setenv("LOTTO_LOGGER_FILE", flags_get_sval(flags, flag_logger_file()),
           true);

    sys_fprintf(stdout, "trace file: %s\n",
                flags_get_sval(flags, flag_input()));

    uint64_t last_clk = cli_trace_last_clk(flags_get_sval(flags, flag_input()));
    uint64_t goal     = flags_get_uval(flags, flag_replay_goal());
    if (goal > last_clk) {
        flags_set_default(flags, flag_replay_goal(), uval(last_clk));
    }

    const char *input_fn = flags_get_sval(flags, flag_input());
    trace_t *rec         = cli_trace_load(input_fn);
    record_t *first      = trace_next(rec, RECORD_START);
    if (first == NULL) {
        sys_fprintf(stderr, "error: empty trace file '%s'\n", input_fn);
        return 1;
    }

    args = record_args(first);
    args_t replay_args = *args;
    char **dynamic_argv = NULL;
    char **original_argv = replay_args.argv;
    execute_resolve_replay_args(&replay_args, flags);
    if (replay_args.argv != original_argv) {
        dynamic_argv = replay_args.argv;
    }

    preload(flags_get_sval(flags, flag_temporary_directory()), verbose,
            !flags_is_on(flags, flag_no_preload()),
            flags_get_sval(flags, flag_memmgr_runtime()),
            flags_get_sval(flags, flag_memmgr_user()));

    trace_advance(rec);
    statemgr_record_unmarshal(trace_next(rec, RECORD_CONFIG));
    trace_destroy(rec);

    envvar_t vars[] = {
        {"LOTTO_REPLAY", .sval = flags_get_sval(flags, flag_input())},
        {NULL}};
    envvar_set(vars, true);

    const char *output_fn = flags_get_sval(flags, flag_output());
    if (output_fn && output_fn[0]) {
        envvar_t vars[] = {{"LOTTO_RECORD", .sval = output_fn}, {NULL}};
        envvar_set(vars, true);
    }
    record_granularities_t record_granularity =
        flags_get_uval(flags, flag_record_granularity());
    char var[RECORD_GRANULARITIES_MAX_LEN];
    record_granularities_str(record_granularity, var);
    setenv("LOTTO_RECORD_GRANULARITY", var, true);
    if (verbose > 0) {
        sys_fprintf(stdout, "[lotto] replaying: ");
        args_print(&replay_args);
        sys_fprintf(stdout, "\n");
    }
    adjust(flags_get_sval(flags, flag_input()));
    int ret = execute(&replay_args, flags, true);
    if (dynamic_argv != NULL) {
        sys_free(dynamic_argv);
    }
    return ret;
}

static flags_t *
_default_flags()
{
    flags_t *flags = flagmgr_flags_alloc();
    flags_cpy(flags, flags_default());
    flags_set_default(flags, flag_output(), sval(""));
    return flags;
}

ON_DRIVER_REGISTER_COMMANDS({
    flag_t sel[] = {flag_input(),
                    flag_output(),
                    flag_verbose(),
                    flag_replay_goal(),
                    flag_temporary_directory(),
                    flag_no_preload(),
                    flag_before_run(),
                    flag_after_run(),
                    flag_logger_file(),
                    0};
    subcmd_register(replay, "replay", "", "Replay a trace", true, sel,
                    _default_flags, SUBCMD_GROUP_TRACE);
})
